File: fpinjava-parent/fpinjava-usingfunctions-exercises/src/main/java/com/fpinjava/functions/exercise02_06/Function.java
package com.fpinjava.functions.exercise02_06;


public interface Function<T, U> {

  U apply(T arg);

  static <T, U, V> Function<Function<U, V>, Function<Function<T, U>, Function<T, V>>> higherCompose() {
    return f -> g -> x -> f.apply(g.apply(x));
  }

  static <T, U, V> Function<Function<T, U>, Function<Function<U, V>, Function<T, V>>> higherAndThen() {
    return f1 -> f2 -> arg -> f2.apply(f1.apply(arg));
  }
}

File: fpinjava-parent/fpinjava-usingfunctions-exercises/src/main/kotlin/com/fpinkotlin/functions/exercise02_06/Function.kt
package com.fpinkotlin.functions.exercise02_06

infix fun <T, U, V> ((U) -> (V)).compose(f: (T) -> (U)) =
		{ arg: T -> this(f(arg)) }

infix fun <T, U, V> ((U) -> (V)).andThen(f: (V) -> (T)) =
		{ arg: U -> f(this(arg)) }
